flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/allocator/StateSizeEstimates.java

```diff
@@ -0,0 +1,129 @@
+                .flatMap(store -> Optional.ofNullable(store.getLatestCheckpoint()))
+                .map(
+                        cp ->
+                                build(
+                                        fromCompletedCheckpoint(cp),
+                                        mapVerticesToOperators(executionGraph)))
+                .orElse(empty());
+    }
+
+    private static StateSizeEstimates build(
+            Map<OperatorID, Long> sizePerOperator,
+            Map<JobVertexID, Set<OperatorID>> verticesToOperators) {
+        Map<JobVertexID, Long> verticesToSizes =
+                verticesToOperators.entrySet().stream()
+                        .collect(
+                                toMap(Map.Entry::getKey, e -> size(e.getValue(), sizePerOperator)));
+        return new StateSizeEstimates(verticesToSizes);
+    }
+
+    private static long size(Set<OperatorID> ids, Map<OperatorID, Long> sizes) {
+        return ids.stream()
+                .mapToLong(key -> sizes.getOrDefault(key, 0L))
+                .boxed()
+                .reduce(Long::sum)
+                .orElse(0L);
+    }
+
+    private static Map<JobVertexID, Set<OperatorID>> mapVerticesToOperators(
+            ExecutionGraph executionGraph) {
+        return executionGraph.getAllVertices().entrySet().stream()
+                .collect(toMap(Map.Entry::getKey, e -> getOperatorIDS(e.getValue())));
+    }
+
+    private static Set<OperatorID> getOperatorIDS(ExecutionJobVertex v) {
+        return v.getOperatorIDs().stream()
+                .map(OperatorIDPair::getGeneratedOperatorID)
+                .collect(Collectors.toSet());
+    }
+
+    private static Map<OperatorID, Long> fromCompletedCheckpoint(CompletedCheckpoint cp) {
+        Stream<Map.Entry<OperatorID, OperatorState>> states =
+                cp.getOperatorStates().entrySet().stream();
+        Map<OperatorID, Long> estimates =
+                states.collect(
+                        toMap(Map.Entry::getKey, e -> estimateKeyGroupStateSize(e.getValue())));
+        return estimates;
+    }
+
+    private static long estimateKeyGroupStateSize(OperatorState state) {
+        Stream<KeyedStateHandle> handles =
+                state.getSubtaskStates().values().stream()
+                        .flatMap(s -> s.getManagedKeyedState().stream());
+        Stream<Tuple2<Long, Integer>> sizeAndCount =
+                handles.map(
+                        h ->
+                                Tuple2.of(
+                                        h.getStateSize(),
+                                        h.getKeyGroupRange().getNumberOfKeyGroups()));
+        Optional<Tuple2<Long, Integer>> totalSizeAndCount =
+                sizeAndCount.reduce(
+                        (left, right) -> Tuple2.of(left.f0 + right.f0, left.f1 + right.f1));
+        Optional<Long> average = totalSizeAndCount.filter(t2 -> t2.f1 > 0).map(t2 -> t2.f0 / t2.f1);
+        return average.orElse(0L);
```

Comment:
I agree that the average isn't a good indicator, but I don't like maximum as well. In fact, I think it's more susceptible to suboptimal decisions.
Percentile should be better than both, I think.

But ideally, we don't aggregate across all subtasks: it's possible to query by `KeyGroupRange` and then compute some aggregate for it (I'd implement this in a separate PR though, to keep this one smaller).

WDYT?

Reply:
> But ideally, we don't aggregate across all subtasks: it's possible to query by KeyGroupRange and then compute some aggregate for it (I'd implement this in a separate PR though, to keep this one smaller).

Agreed, we shouldn't further evolve this PR in this regard.

> In fact, I think it's more susceptible to suboptimal decisions.

Can you expand on that? My feeling is that there few extreme cases:
1) one large state + one small state (this is IMO quite common, e.g. buffer + watermark hold)
2) multiple with the approximately same size

Assumption: it's better to move a smaller state than a big one;

For 1) Max performs better because the small state would average the big one out, even though the price of moving the small one is insignificant;

For 2) both would perform approximately the same

Are there any other cases you have in mind? My problem with percentiles is that it'd actually ignore the worst cases that are most costly to move/split/merge.

Reply:
I'm wondering whether something along these lines holds (for RocksDB): "The rescaling of the Flink job is as slow as the rescaling of the single largest state it contains."

Reply:
I've created FLINK-31261 for state size estimation.